File: ultra-analytics/frontend/src/subscription/features.ts
/** Central feature → tier map. Keep in sync with backend ``subscription/features.py``. */

export type SubscriptionTier = "guest" | "free" | "pro";

export type ProFeature = "planning" | "verify" | "rideMode" | "gpxExport";

export const PRO_FEATURES: readonly ProFeature[] = [
  "planning",
  "verify",
  "rideMode",
  "gpxExport",
] as const;

export function normalizeTier(raw: unknown): SubscriptionTier {
  const v = String(raw ?? "guest").trim().toLowerCase();
  if (v === "pro" || v === "free" || v === "guest") return v;
  return "guest";
}

/** Tier for a session user — logged-in defaults to free; null user is guest. */
export function tierFromUser(user: { subscriptionTier?: string | null } | null | undefined): SubscriptionTier {
  if (!user) return "guest";
  const t = normalizeTier(user.subscriptionTier ?? "free");
  return t === "guest" ? "free" : t;
}

export function canAccess(
  feature: ProFeature | string,
  tier: SubscriptionTier | string | null | undefined,
): boolean {
  const t = normalizeTier(tier ?? "guest");
  if (!(PRO_FEATURES as readonly string[]).includes(feature)) return false;
  return t === "pro";
}

export function canAccessUser(
  feature: ProFeature | string,
  user: { subscriptionTier?: string | null } | null | undefined,
): boolean {
  return canAccess(feature, tierFromUser(user));
}

/** Planned GPX import — Pro sub or an unused Race Pass credit. */
export function canImportPlannedRoute(
  user: { subscriptionTier?: string | null; racePassCredits?: number | null; billing?: { racePassCredits?: number } } | null | undefined,
): boolean {
  if (canAccessUser("planning", user)) return true;
  const credits = user?.racePassCredits ?? user?.billing?.racePassCredits ?? 0;
  return Number(credits) > 0;
}

export function tierLabel(tier: SubscriptionTier | string | null | undefined): string {
  const t = normalizeTier(tier ?? "guest");
  if (t === "pro") return "Pro";
  if (t === "free") return "Free";
  return "Guest";
}

File: ultra-analytics/backend/app/subscription/features.py
"""Central feature → tier map. Keep in sync with frontend ``subscription/features.ts``."""

from __future__ import annotations

from .tiers import TIER_PRO, normalize_tier

# Stable feature ids used by gates (API deps + UI).
FEATURE_PLANNING = "planning"
FEATURE_VERIFY = "verify"
FEATURE_RIDE_MODE = "rideMode"
FEATURE_GPX_EXPORT = "gpxExport"

# Free-tier product areas (documented; no gate — auth alone is enough):
# library, analytics, certificates, ultras, rideHistory

PRO_FEATURES = frozenset(
    {
        FEATURE_PLANNING,
        FEATURE_VERIFY,
        FEATURE_RIDE_MODE,
        FEATURE_GPX_EXPORT,
    }
)

FEATURES = frozenset(PRO_FEATURES)


def can_access(tier: object, feature: str) -> bool:
    """Return True when ``tier`` may use ``feature``.

    Guest (or missing) never passes Pro features. Free never passes Pro.
    Unknown features default to False (fail closed).
    """
    t = normalize_tier(tier, default="guest")
    if feature not in FEATURES:
        return False
    if feature in PRO_FEATURES:
        return t == TIER_PRO
    return t in ("free", "pro")
